src/github.com/getlantern/flashlight/client/client.go

```diff
@@ -39,66 +39,90 @@ type Client struct {
 	priorCfg        *ClientConfig
 	priorTrustedCAs *x509.CertPool
 	cfgMutex        sync.RWMutex
-	balCh           chan *balancer.Balancer
-	balInitialized  bool
-	rpCh            chan *httputil.ReverseProxy
-	rpInitialized   bool
-	hqfd            fronted.Dialer
-	l               net.Listener
+
+	// Balanced CONNECT dialers.
+	balChMu        sync.Mutex
+	balCh          chan *balancer.Balancer
+	balInitialized bool
+
+	// Reverse HTTP proxies.
+	rpChMu        sync.Mutex
+	rpCh          chan *httputil.ReverseProxy
+	rpInitialized bool
+
+	hqfd fronted.Dialer
+	l    net.Listener
 }
 
-// ListenAndServe makes the client listen for HTTP connections.
-// onListening is a callback that gets invoked as soon as the server is
-// accepting TCP connections.
-func (client *Client) ListenAndServe(onListening func()) error {
-	l, err := net.Listen("tcp", client.Addr)
-	if err != nil {
-		return fmt.Errorf("Client unable to listen at %v: %v", client.Addr, err)
+// ListenAndServe makes the client listen for HTTP connections.  onListeningFn
+// is a callback that gets invoked as soon as the server is accepting TCP
+// connections.
+func (client *Client) ListenAndServe(onListeningFn func()) (err error) {
+	var l net.Listener
+
+	// Attempting to bind a port on the given address.
+	if l, err = net.Listen("tcp", client.Addr); err != nil {
+		return fmt.Errorf("Client proxy was unable to listen at %s: %q", client.Addr, err)
 	}
-	client.l = l
 
-	onListening()
+	// We opened the port successfully at this point, copying listener to our
+	// client and executing the passed callback.
+	client.l = l
+	onListeningFn()
 
+	// Creating an HTTP server.
```

Comment:
TBH all the comments in this body, especially this one, are just restating what the code says.  I find that more distracting than helpful.